### ISSUE TITLE
Router 3.2.0 - Providers instantiated twice

### ISSUE THREAD

**Issue by slubowsky** [angular/angular#12869]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
After upgrading to router 3.2.0, services being provided by component module are being instantiated twice. Once while resolver is being run (last place in router is PreActivation.prototype.getToken)  and then again later during activation (last place in router is RouterOutlet.prototype.activate).
Upgrading rest of Angular to 2.2.0 and leaving router at 3.1.1 it works as it used to (and looking in debugger I see service is only instantiated from PreActivation.prototype.getToken)

**Expected behavior**

Providers should only be instantiated once - as they are when using router version 3.1.1

**Minimal reproduction of the problem with instructions**

Haven't yet reproduced in a plnkr but my use case has a lazy loaded component that itself has sub components (not lazy loaded). The sub component has a resolver that uses a service, and the sub component itself uses the service. The service is included in the lazy loaded components module providers list and is being instantiated twice. Once apparently while router is working on resolve and once when component is being activated.

**What is the motivation / use case for changing the behavior?**

Its just wrong, and my service fetches data on initialization so extra instance causes extra fetch. This does not happen with router 3.1.1 (never tried 3.1.2 as deactivate guards don't work correctly in that version)

**Please tell us about your environment:**

Windows 10, Visual Studio Code, angular-cli: 1.0.0-beta.19-3 ng serve

* **Angular version:** 2.0.X

2.2.0

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

 Only tried in Chrome

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
Typescript 2.0.9

* **Node (for AoT issues):** `node --version` =

**Comment by DzmitryShylovich** [comment]
Pls add a repro in plunkr.

**Comment by slubowsky** [comment]
Modified an existing simple repo I made for a different issue - https://github.com/slubowsky/routing-angular-cli. Run the app (ng serve) and click the "lazy" link. You will see the FooService constructor is called twice. It shouldn't be.
Change router to 3.1.1 and rerun. The constructor is only called once.

**Comment by vicb** [comment]
Could you reduce your repo to only the things required for demonstrating the issue

**Comment by slubowsky** [comment]
Ok just deleted most of what was there (there wasn't much). Its pretty bare bones now. A top level app component and one lazy loaded child. The child is the default route and is loaded on startup. You can see my console.log message from the service constructor twice. Change to router 3.1.1 it and only appears once.

**Comment by thaoula** [comment]
It seems to running providers starting with twice, then 3 times, then 4 times etc ... each time you go into a Lazy Loaded route.

I started seeing this effect when watch network requests as soon as updated to 3.2.0 of the router. 3.1.0 version has no problems.

**Comment by serhiisol** [comment]
I got the same problem, module constructor executed twice for lazy loaded route

**Comment by antonybudianto** [comment]
It's [fixed](https://github.com/angular/angular/pull/10707) right? No patch release until now?

**Comment by httpdigest** [comment]
Having the exact same problem as described by [@serhiisol](https://github.com/angular/angular/issues/12869#issuecomment-261217933).

**Comment by thaoula** [comment]
Hi Guys,

We need to start using some new features on FormGroup (2.2.0) but are blocked by this issue.

Can someone from the Angular team advise when this will be resolved. The issue appears to exist in 2.2.3 and the latest 2.3.0 beta.

- Are there any known workarounds?
- Is it something we need to change in our code?

Regards,
Tarek

**Comment by httpdigest** [comment]
Attached is a angular-cli project which exactly reproduces the bug as I see it.
It has a lazy route which, when loaded, leads to the module being loaded _twice_, as is shown by the console.log() in the `src/app/lazy.module.ts` file:
[router-bug.zip](https://github.com/angular/angular/files/619860/router-bug.zip)

**Comment by JBusch** [comment]
Same problem here.

**Comment by antonybudianto** [comment]
This kind of bug should be put to high priority I think, need patch update soon

**Comment by Dunos** [comment]
I thought I was the only one with this issue but clearly I'm not, let's hope they fix it soon :/

**Comment by vsavkin** [comment]
This is why it is happening.

* Say we have two modules Parent and Child, where Child is loaded lazily. 
* Say we have two components ParentCmp1 and ParentCmp2 defined in Parent, and ChildCmp1 defined in Child.
* Say typically the components form the following tree when instantiated: ParentCmp1 => ParentCmp2 => ChildCmp1.
* Finally, say ChildCmp1 injects ParentCmp2.

Since we run guards and resolvers before instantiating components (this is one of the guarantees the router provides to make component instantiation atomic), we can only constructor the following injector tree:

```
Injector for Parent => Injector for Child.
```

When instantiating components, the injector tree has to look like this:

```
Injector for Parent => Injector for ParentCmp1 => Injector for ParentCmp2 => Injector for Child => Injector for ChildCmp1
```

Because we have two injector trees, we instantiate the module twice. One module cannot be a node in two trees, it cannot be used in two contexts.

@vicb what do you think is a good solution? we can add a new property something like loadGuards in addition to loadChildren, so it's possible to provision two modules (one for preActivation, and the other one for activation).

**Comment by thaoula** [comment]
Hi Victor,

Thanks for the info.

Do you know what was changed in 2.2.0 to break the previous experience? In 2.1.0 this was not a issue?

Also, what is really the primary reason behind creating different DI contexts for lazy loaded modules. As they seem to be creating more friction than benefits. 

I would imagine that in a large majority of applications, the authors of the lazy loaded modules are part of the same project. So the modules that are being lazy loaded are known to them.

The issue of overwriting providers can be managed by communication within the team and some sort of angular warning. Many of us came from Angular 1 where we only had a single DI context and we were mostly ok.

IMHO, I would much prefer having a setting that enables the use of a single DI context for all modules including lazy loaded modules. 

This would eliminate the need for -
- ForRoot
- ForChild
- loadGuards option (indicates detailed knowledge of the lazy loaded module anyway)

Regards,
Tarek

**Comment by chrissena** [comment]
Any news on this issue?

**Comment by manofearth** [comment]
Any news? I use ngrx/effects and such behavior doubles my effect-observables and I have two same working effects instead of only one, and reducers fires twice with same action payload.

**Comment by ciesielskico** [comment]
@manofearth Use an older version of angular/router until this issue is closed, I was having the same problem.

**Comment by michalboska** [comment]
@manofearth I am using 3.1.1 for the time being

**Comment by manofearth** [comment]
@ciesielskico @michalboska I found another workaround: include all effects in root module

**Comment by michalboska** [comment]
@manofearth For me this is unfortunately a no-go, my app is pretty large and i really need to split it into modules. They are actually lazy loaded too.

**Comment by nickiewooster** [comment]
Any ETA for this ?

**Comment by rhessus** [comment]
I confirm, there is not this issue with the 3.1.x version of the router. When i update to >= 3.2.0, the problem appears.

**Comment by montella1507** [comment]
+1 i think this is critical issue because it is the cause of problems with every non-stateless services.. like ngrx/store, effecs,.. translate etc.

**Comment by thaoula** [comment]
I have tried tweeting members of the team with no response. I just cannot believe this bug is still not resolved. 

This is pretty much now a show stopper for us now, we cannot upgrade beyond Angular 2.1. 

Are people just not using lazy loading?

**Comment by robwormald** [comment]
Update: we had a discussion / design review today on this, and think we have a workable solution to solve it. Just verifying the semantics are right and we'll land a PR.

**Comment by lreading** [comment]
Has anyone experienced this without lazy loading?  The symptoms are all the same as what's reported here, but if seems as though everyone here is lazy loading where as I am not.

**Comment by akalitenya** [comment]
@lreading yes, when i have accidentally added same line 'EffectsModule.run(AppEffects)' (to run ngrx effects) twice in app module's @NgModule imports.

**Comment by CraigAP** [comment]
I'm trying to use a service to share data needed by multiple components.  I set a break on the service constructor and it breaks every time it is accessed.  Documetation says services are singletons instantiated once.  Am I wrong to store app-wide data in a service?

**Comment by mhevery** [comment]
We have a [design doc](https://docs.google.com/document/d/1OEUIwc-s69l1o97K0wBd_-Lth5BBxir1KuCRWklTlI4) on this issue, but we will not be able to get to it. Would someone in the community like to pick it up?

/cc @DzmitryShylovich Would you be up for this? I am happy to provide technical support.

**Comment by DzmitryShylovich** [comment]
@mhevery I'll take a look

**Comment by mxchange** [comment]
I'm having the same issue when I `EffectsModule.run(AppEffects)` in a lazy loaded module,
And `@angular/router version `3.1.1` like @michalboska suggested does resolve this. (One problem with that is that I don't know where to find version numbers of that, because it is packaged with angular2? And any other `@angular/` module I think will override the version number?)

It will fire the action twice after loading the lazy route, but when I load the effect in the root module with either `.run` or `.runAfterBootstrap`, the `fulfilled` action will fire only once,
And every time I go back to the same lazy loaded route, it will add `1x` to the actions dispatched, although it will only `fetch` once:

![ngrx effects router bug](https://cloud.githubusercontent.com/assets/24876534/22460631/1095872a-e7a6-11e6-811f-802142a6306b.png)


My effects module, nothing special: 

```

import 'rxjs/add/observable/of';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/switchMap';

import { Injectable } from '@angular/core';
import { Actions, Effect } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { ApiService } from '@core/shared/services/api';
import { LogsActions } from '@core/actions/logs';


@Injectable()
export class LogEffects {
  constructor(
    private actions$: Actions,
    private api: ApiService,
    private logActions: LogsActions
  ) {}

  @Effect()
  fetchLogs$ = this.actions$
    .ofType(LogsActions.FETCH_LIVE_LOGS)
    .switchMap( () => {
      console.info('Effect: Logs');
      return this.api.fetchLogs()
        .map(res => (this.logActions.fetchLogsFulfilled(res)))
        .catch(error => ( Observable.of( this.logActions.fetchLogsFailed(error) )));
      }
    );
}
```

**Comment by karolmie1** [comment]
@mxchange Yup, you are describing same issue everyone has. Yeah, router has different numeration from rest of angular components (namely: 3.x.x instead of 2.x.x. - it's because they've changed router heavily half way, and had to increase major version to comply with semver). Soon the problem will be fixed, as angular4 will be released specially to avoid this types of confusion. 
For now you just have to stick with old depenencies, for example: 

`
    "@angular/platform-browser": "2.1.0",
    "@angular/platform-browser-dynamic": "2.1.0",
    "@angular/router": "3.1.0",
`

**Comment by mxchange** [comment]
@karolmie1 thanks, angular4 would be great, i've tried it out, but the last `beta.5` (for /router as well) still has the same issue of firing twice.

As I understand, this specific issue should be fixed when it's out of beta?

**Comment by karolmie1** [comment]
@mxchange : glorious DzmitryShylovich is taking care of it now. Unfortunately it's not high priority and it's planned to be fixed before angular4 release. You have to stick to old version for now.

**Comment by k-schneider** [comment]
Is there an ETA for this?  The PR for the 2.x fix has been sitting dormant for ~10 days awaiting approval.  I know a lot of people are waiting on this fix to upgrade (myself included). Thanks!

**Comment by feng-xiao** [comment]
any updates on this issue? for me, the service (_CustomerStateDataService_) injected in lazy loading module's resolver constructor get instantiated twice: first instance created by the resolver service, second instance is instantiated by first component in the lazy loading module that also injecting the same service (_CustomerStateDataService_) used by resolver service. I have to define the service in first early loaded parent module (_RootModule_) to fix the issue.

**Working (defined in first early-loaded parent module; NgModule found for injectiong the service - CustomerStateDataService):** 
![image](https://cloud.githubusercontent.com/assets/3407775/23036479/fa12ce38-f450-11e6-86ed-a78c903b29ad.png)


**The way I want, but not working (defined in lazy-loaded module; no NgModule found for injecting the service - CustomerStateDataService):** 
![image](https://cloud.githubusercontent.com/assets/3407775/23036223/01c8c1b0-f450-11e6-933b-2d8648ad427d.png)

**Comment by karptonite** [comment]
@feng-xiao There is a pull request out to fix this issue in the 2.4 branch:
https://github.com/angular/angular/pull/14201
I'm not sure what the timeline is for merging it, but that is the one to watch if you are on 2.4.

**Comment by feng-xiao** [comment]
@karptonite thanks! I'll take a look.

**Comment by thaoula** [comment]
Guys, 

I have tweeted the team again to ask when this will the fix will merged as I don't know how else to communicate with them.

Regards,
Tarek

**Comment by DzmitryShylovich** [comment]
@thaoula the fix will be available in 4.0 final (and maybe in 2.x)

**Comment by thaoula** [comment]
Hi DzmitryShylovich,

Firstly, many thanks for your efforts on creating a fix for 2.x and for 4.x.

Are you saying we are stuck on 2.1 and 3.1 until version 4 come out? 

Will a fix be in any 4.x beta's or anything we can use now?

As it stand now, if we try and upgrade our application it will collapse due to this issue. 

Ngrx effects, Http calls will run 1 x for each route... if a route has 3 children routes things will run 4 times. If you go in an out of a lazy loaded module, it will run 2 x route + child routes, this increases each time you reload a lazy loaded module.

I am sure it effects everyone. People may not notice if they are just creating hello world apps.

This is a serious bug, not a nice to have feature.

Regards,
Tarek

**Comment by DzmitryShylovich** [comment]
> Are you saying we are stuck on 2.1 and 3.1 until version 4 come out?

don't use lazy loaded modules for now. It should be pretty straightforward to switch in the future.

**Comment by kdorion** [comment]
4.0 final should be out at the beginning of march, so its not that bad considering we've been waiting since december. We are still using 2.1 because we need lazy loaded modules for permission issues.

**Comment by DzmitryShylovich** [comment]
> 4.0 final should be out at the beginning of march

don't think so :) ~mid April I guess

**Comment by kdorion** [comment]
Well, my comment was according to when angular 4 was first announced :p

**Comment by thaoula** [comment]
Hi Kdorion/DzmitryShylovich,

We are in a similar boat with Lazy loading to enforce permissions. So it is not easy to just switch.

DzmitryShylovich, I am pretty sure you fixed this issue long time ago. Why can't it be applied to 2.4.x?

Regards,
Tarek

**Comment by DzmitryShylovich** [comment]
>  Why can't it be applied to 2.4.x?

idk I'm not a team member ¯\_(ツ)_/¯

**Comment by bevrard** [comment]
Is there a way to influence the Angular team to integrated in 2.4.x? :) As @thaoula said, we are not just creating hello world apps.  By the way thanks @DzmitryShylovich for the fix! 👍

**Comment by DzmitryShylovich** [comment]
> Is there a way to influence the Angular team to integrated in 2.4.x?

they are busy on 4.0 rc release. After that we will get some feedback I guess.

**Comment by mhevery** [comment]
We are very aware of this issue but unfortunately have higher priority
things on our plate. Can't promise but we would like to get this done for
4.0.0.

**Comment by tomwanzek** [comment]
@mhevery As always, thanks to yourself and the entire Angular team for all the hard work.

Just for clarity, does

> Can't promise but we would like to get this done for 4.0.0.

imply that a fix may not be available as of at least 4.0.0? And by implication, that fixing the regression in 2.4.x is completely out of the question?

As mentioned above by others, the issue _currently_ already affects **ngrx effects**. And while it may be a speculative stretch, at face value, it seems to impact the next phase of **ngrx store** to include the long-awaited full support for lazy-loading. See [ngrx store Issue 211](https://github.com/ngrx/store/issues/211).

My apologies in advance, if I am misinterpreting something.

**Comment by abidmix** [comment]
As of now I have resorted to directly converting @DzmitryShylovich  fix to javascript and patching those lines of code into the @angular/router/src/router_config_loader.js. Because we will be stuck for a while using angular 2.4.5 The only downside of  this is that  our build process cleans the node _modules first and I have to manually overwrite this file. @mhvery can you see any issue of running this patch in production since we have to wait for 4.0.0. Unfortunately we can't stop using lazy modules now..our app is so big and heavily relies on them.

**Comment by mxchange** [comment]
Still not fixed in 4.0.0.rc1, anyone @googlebot, @google-admin responsible for prioritizing pull requests? https://github.com/angular/angular/pull/13593

**Comment by mhevery** [comment]
Just an FYI, @vicb  is working on this issue now. We are cautiously optimistic, about it making into v4.0.0

**Comment by Matmo10** [comment]
Does this mean we're on target for v4? 😄

**Comment by DzmitryShylovich** [comment]
next rc

**Comment by brandonroberts** [comment]
Thanks @mhevery @vicb and the rest of the team for getting this fix in the V4 👍

**Comment by rhessus** [comment]
Thanks a lot for your hard work !!!

**Comment by mxchange** [comment]
Can anyone confirm this is fixed?

**Comment by slubowsky** [comment]
@mxchange Yes, It is fixed. In my case, same project that used to get multiple instantiations of singleton providers no longer does with Angular 4.0.

**Comment by ldex** [comment]
Fixed for me as well since 4.0!

**Comment by JohnnyClutch** [comment]
This is not fixed.  I have upgraded to angular 4.1.3.  

My core module loads our shared library module and imports the permissions service from it.  It is able to interact successfully with it and the data cached / stored within it.  

Through a lazy loaded route, my ribbons component utilizes a directive imported from the same shared module that gets the permissions service injected into it.  The version of that service is a fresh copy so the data cached in the original load of the service is missing.  I know it is a copy because I set breakpoints on the constructor of the permissions service, and they hit twice.

I'd be happy to set up a plunkr if anyone would like to see it.

**Comment by chuckjaz** [comment]
@JohnnyClutch This might be a different issue that references this one. Can you create a separate bug and attach the plunkr?

**Comment by pgrm** [comment]
@JohnnyClutch @chuckjaz - I see the same issue with latest Angular Router when enabling `preloadingStrategy: PreloadAllModules` - when I disable preloading than it works as expected.

**Comment by craigsnyders** [comment]
@pgrm's solution worked for me, after noticing that I was double-declaring the service!

I guess we need another bug raised for `preloadingStrategy: PreloadAllModules`

**Comment by jonathanborges** [comment]
Does'not work for me with ```PreloadAllModules``` disabled.

Using ```loadChildren``` and ```forChild``` routes on child modules.

Angular 4.3.4

**Comment by rlegrand** [comment]
I maybe miss something, but for me it's not fixed in all cases.
In the app below: 
* the app has a dependency to a core module
* the app lazily loads a "lazy" module
* the lazy module has a dependency to the core module
* both the app module and the lazy module use the singleton service provided in the core module
=> The service is instantiaited twice.

https://stackblitz.com/edit/angular-7sqt76
Run the app and look at logs.

If there is something wrong in my declarations which explains that, please tell me what.
IF not, do you have a workaround in mind while I need my singletons to be singletons.. It's the basis !

**Comment by mlc-mlapis** [comment]
@rlegrand ... I can't imagine why you should provide the singleton in the core module and not directly in the app module but if you want to do it in any reason ... you need to use the following syntax ... https://stackblitz.com/edit/angular-7sqt76-2hidqf?file=app%2Fmodules%2Fcore%2Fcore.module.ts

**Comment by ldex** [comment]
Hi @rlegrand, what you get is the expected behavior.
To achieve your scenario you have to use the for root() technic.
Please have a look at the doc: https://angular.io/guide/singleton-services

**Comment by rlegrand** [comment]
Hi,
Thank you guyz.
Without this answer, I would probably have used a static attribute in my service to be sure to manage only one instance"manually".

@mlc-mlapis  I called it core module and includes it in the app for the example, in fact I just mean "I have a module which provides services"

Thank you.

**Comment by mlc-mlapis** [comment]
@rlegrand ... ah, ok, I understand ... but I still think that the idea of `I have a module which provides services` is more recombined as it should be ... especially with just singletons providers ... so I recommend to include them directly to `providers` section of the app module.

With other parts ... like components, pipes, ... the idea of `core module` is still fine.

**Comment by sijucm** [comment]
I still have the issue with latest version. I declared a CoreModule with forRoot as shown in the https://angular.io/guide/singleton-services and included that in the AppModule. But still the supposedly Singleton service is created every time I navigate to routes that injects them (uses them). I'm going crazy searching for a solution because I thought this was basic functionality and checking and rechecking that I did not make a mistake. I'm following text book instructions. The comments above asking why a CoreModule is created is irrelevant because the CoreModule solution is tried as per instruction. Otherwise I do not understand why the providers cannot be declared directly in the AppModule.

**Comment by ghetolay** [comment]
@sijucm You're providing `UserService` inside `CustomerComponents` so the service follow the lifecycle of the component: it is created each time a customer component is created and destroyed when the component is destroyed.

Issues tracker is not meant to provide support so please use stackoverflow or gitter next time you need help.

**Comment by gruppjo** [comment]
There is an answer on stackoverflow to prevent services in a shared module to be instantiated multiple times that worked for me:

Implement `forRoot()` on that shared module and only add providers there. In app module import shared module with `forRoot()` and in all feature modules use the `SharedModule` without `forRoot()`.

https://stackoverflow.com/questions/41290275/angular2-lazy-loading-service-being-instantiated-twice

**Comment by lucasbasquerotto** [comment]
@gruppjo Is there some recommended approach to handle cases where the app is not small and the size of the singleton providers is not trivial to load all of them in the initial load?

In most cases I see, providers are supposed to be singletons (except for some specific cases). For small apps this approach may seem fine, but as the app grows you end up declaring a lot of providers in `AppModule`, making the initial load size larger and consequently slower. 

Also, this doesn't seem a good approach with libraries because you would have to load all the libraries providers (or most of them) in the initial load (in AppModule).

Is there some recommendation regarding that? Like defining one global provider that store the state of other providers and import it in `AppModule` and import all other providers only where they are used (accessing the global provider to retrieve and define state, instead of relying in itself). Or something along these lines?

Otherwise it might become difficult to maintain the performance of angular apps as the app size grows.

**Comment by anth-git** [comment]
I agree, it is very misleading. After reading documentation I had clear conviction that services declared in modules are singletons, so I don't understand why it behaves differently for lazy loaded modules. At the very least this behavior should be described in documentation. Admittedly there is this statement in docs:

"Beginning with Angular 6.0, the preferred way to create a singleton services is to specify on the service that it should be provided in the application root"

but, as far as I'm concerned, for now, it is the only way, not the 'preffered way'.

**Comment by ilyes-hamrouni** [comment]
I have the problem but i am not working with any modules except the app.module.ts, please check my case.
https://stackoverflow.com/questions/54391955/angular-service-have-multiple-instances-instead-of-singleton/54392133?noredirect=1#comment95596734_54392133
